### ISSUE TITLE
Problems with the cell definition and setting parameters

### ISSUE THREAD

**Issue by dariosannino** [dbbs-lab/arborize#28]
Hi, I am having some problems in building a cell with appropriate parameters defined in the definition.
The definition is the following:
```
from arborize import define_model
from arborize import bsb_schematic
from bsb.morphologies import Morphology
from arborize import neuron_build
import numpy as np
from patch import p

definitionGrCWT = define_model({
    "synapse_types": {
        "AMPA": {
            "mechanism": ("AMPA", "granule"),
            "attributes": {
                "tau_facil": 5,
                "tau_rec": 8,
                "tau_1": 1,
                "gmax": 1400,
                "U": 0.43,
            },
        },
        "NMDA": {
            "mechanism": ("NMDA", "granule"),
            "attributes": {
                "tau_facil": 5,
                "tau_rec": 8,
                "tau_1": 1,
                "gmax": 23500,
                "U": 0.43,
            },
        },
        "GABA": {"mechanism": ("GABA", "granule"), "attributes": {"U": 0.35}}
    },
    "cable_types": {

        "soma": {
            "cable": {"Ra": 100, "cm": 2},
            "ions": {"k": {"e": -80.993}, "ca": {"e": 137.5}},      #ek might be -88, before was -80.993
            "mechanisms": {
                "Leak": {"e": -60, "gmax": 0.00029038073716},
                "Kv3_4": {"gkbar": 0.00076192450952},
                "Kv4_3": {"gkbar": 0.00281496839061},
                "Kir2_3": {"gkbar": 0.00074725514702},
                "Ca": {"gcabar": 0.00060938071784},
                "Kv1_1": {"gbar": 0.00569738264555},
                "Kv1_5": {"gKur": 0.00083407556714},
                "Kv2_2": {"gKv2_2bar": 1.203410852e-05},
                ("cdp5", "CR"): {},
            },
        },
        "dendrites": {
            "cable": {"Ra": 100, "cm": 2.5},
            "ions": {"k": {"e": -80.993}, "ca": {"e": 137.5}},      # ek might be -88
            "mechanisms": {
                "Leak": {"e": -60, "gmax": 0.00025029700737},
                ("Leak", "GABA"): {"ggaba" : 0.0},  #Set to 0
                "Ca": {"gcabar": 0.00500128008459},
                "Kca1_1": {"gbar": 0.01001807454651},
                "Kv1_1": {"gbar": 0.00381819207934},
                ("cdp5", "CR"): {},
            },
        },
        "axon": {"cable": {}, "ions": {}, "mechanisms": {}},
        "ascending_axon": {
            "cable": {"Ra": 100, "cm": 1},
            "ions": {"na": {"e": 87.39}, "k": {"e": -80.993}, "ca": {"e": 137.5}},      # ek might be -88
            "mechanisms": {
                ("Na", "granule_cell"): {"gnabar": 0.02630163681502},
                "Kv3_4": {"gkbar": 0.00237386061632},
                "Leak": {"e": -60, "gmax": 9.364092125e-05},
                "Ca": {"gcabar": 0.00068197420273},
                ("cdp5", "CR"): {},
            },
        },
        "axon_initial_segment": {
            "cable": {"Ra": 100, "cm": 1},
            "ions": {"na": {"e": 87.39}, "k": {"e": -80.993}, "ca": {"e": 137.5}},        # ek might be -88
            "mechanisms": {
                ("Na", "granule_cell_FHF"): {"gnabar": 1.28725006737226},
                "Kv3_4": {"gkbar": 0.00649595340654},
                "Leak": {"e": -60, "gmax": 0.00029276697557},
                "Ca": {"gcabar": 0.00031198539472},
                "Km": {"gkbar": 0.00056671971737},
                ("cdp5", "CR"): {},
            },
        },
        "axon_hillock": {
            "cable": {"Ra": 100, "cm": 2},
            "ions": {"na": {"e": 87.39}, "k": {"e": -80.993}, "ca": {"e": 137.5}},                      # ek might be -88
            "mechanisms": {
                "Leak": {"e": -60, "gmax": 0.0003695818972},
                ("Na", "granule_cell_FHF"): {"gnabar": 0.00928805851462},
                "Kv3_4": {"gkbar": 0.02037346310915},
                "Ca": {"gcabar": 0.00057726155447},
                ("cdp5", "CR"): {},
            },
        },
        "parallel_fiber": {
            "cable": {"Ra": 100, "cm": 1},
            "ions": {"na": {"e": 87.39}, "k": {"e": -80.993}, "ca": {"e": 137.5}},                      # ek might be -88
            "mechanisms": {
                ("Na", "granule_cell"): {"gnabar": 0.01771848449261},
                "Kv3_4": {"gkbar": 0.00817568047037},
                "Leak": {"e": -60, "gmax": 3.5301616e-07},
                "Ca": {"gcabar": 0.0002085683353},
                ("cdp5", "CR"): {},
            },
        },
    },
})


tagsGrC = {
    16: ["axon", "axon_hillock"],
    17: ["axon", "axon_initial_segment"],
    18: ["axon", "ascending_axon"],
    19: ["axon", "parallel_fiber"],
}
```

Then I build the cell:
```
morpho = Morphology.from_swc("Morphologies/GranuleCell.swc", tags=tagsGrC)
schematic = bsb_schematic(morpho, definitionGrCWT)
cell = neuron_build(schematic)
```

Then I try to check if everything is ok, for instance at the level of the soma:
```
cell.soma[0].psection()
```

And I obtain the following output:
```
{'point_processes': {}, 'density_mechs': {'glia__dbbs_mod_collection__Kir2_3__0': {'Aalpha_d': [0.13289], 'Kalpha_d': [-24.3902], 'V0alpha_d': [-83.94], 'Abeta_d': [0.16994], 'Kbeta_d': [35.714], 'V0beta_d': [-83.94], 'gkbar': [0.00074725514702], 'ik': [0.0], 'd_inf': [0.0], 'tau_d': [0.0], 'g': [0.0], 'alpha_d': [0.0], 'beta_d': [0.0], 'd': [0.0]}, 'glia__dbbs_mod_collection__Kv1_1__0': {'gateCurrent': [0.0], 'gbar': [0.00569738264555], 'gunit': [16.0], 'ik': [0.0], 'i': [0.0], 'igate': [0.0], 'g': [0.0], 'nc': [0.0], 'ninf': [0.0], 'taun': [0.0], 'n': [0.0]}, 'glia__dbbs_mod_collection__cdp5__CR': {'Nannuli': [10.9495], 'cainull': [4.5e-05], 'Buffnull2': [60.9091], 'rf3': [0.1435], 'rf4': [0.0014], 'CRnull': [0.9], 'TotalPump': [1e-09], 'ica_pmp': [0.0], 'vrat': [0.0], 'ca': [0.0], 'mg': [0.0], 'Buff1': [0.0], 'Buff1_ca': [0.0], 'Buff2': [0.0], 'Buff2_ca': [0.0], 'BTC': [0.0], 'BTC_ca': [0.0], 'DMNPE': [0.0], 'DMNPE_ca': [0.0], 'CR': [0.0], 'CR_1C_0N': [0.0], 'CR_2C_0N': [0.0], 'CR_2C_1N': [0.0], 'CR_1C_1N': [0.0], 'CR_0C_1N': [0.0], 'CR_0C_2N': [0.0], 'CR_1C_2N': [0.0], 'CR_2C_2N': [0.0], 'CR_1V': [0.0], 'pump': [0.0], 'pumpca': [0.0]}, 'glia__dbbs_mod_collection__Ca__granule_cell': {'Aalpha_s': [0.04944], 'Kalpha_s': [15.873], 'V0alpha_s': [-29.06], 'Abeta_s': [0.08298], 'Kbeta_s': [-25.641], 'V0beta_s': [-18.66], 'Aalpha_u': [0.0013], 'Kalpha_u': [-18.183], 'V0alpha_u': [-48.0], 'Abeta_u': [0.0013], 'Kbeta_u': [83.33], 'V0beta_u': [-48.0], 'gcabar': [0.00060938071784], 'ica': [0.0], 's_inf': [0.0], 'u_inf': [0.0], 'tau_s': [0.0], 'tau_u': [0.0], 'g': [0.0], 'alpha_s': [0.0], 'beta_s': [0.0], 'alpha_u': [0.0], 'beta_u': [0.0], 's': [0.0], 'u': [0.0]}, 'glia__dbbs_mod_collection__Kv2_2__0': {'gKv2_2bar': [1.203410852e-05], 'BBiD': [10.0], 'ik': [0.0], 'gKv2_2': [0.0], 'm': [0.0], 'h': [0.0]}, 'glia__dbbs_mod_collection__Kv4_3__0': {'Aalpha_a': [0.8147], 'Kalpha_a': [-23.3271], 'V0alpha_a': [-9.17203], 'Abeta_a': [0.1655], 'Kbeta_a': [19.4718], 'V0beta_a': [-18.2791], 'Aalpha_b': [0.0368], 'Kalpha_b': [12.8433], 'V0alpha_b': [-111.332], 'Abeta_b': [0.0345], 'Kbeta_b': [-8.90123], 'V0beta_b': [-49.9537], 'V0_ainf': [-38.0], 'K_ainf': [-17.0], 'V0_binf': [-78.8], 'K_binf': [8.4], 'gkbar': [0.00281496839061], 'ik': [0.0], 'a_inf': [0.0], 'b_inf': [0.0], 'tau_a': [0.0], 'tau_b': [0.0], 'g': [0.0], 'alpha_a': [0.0], 'beta_a': [0.0], 'alpha_b': [0.0], 'beta_b': [0.0], 'a': [0.0], 'b': [0.0]}, 'glia__dbbs_mod_collection__Leak__0': {'gmax': [0.00029038073716], 'e': [-60.0], 'il': [0.0]}, 'glia__dbbs_mod_collection__Kv1_5__0': {'gKur': [0.00083407556714], 'Tauact': [1.0], 'Tauinactf': [1.0], 'Tauinacts': [1.0], 'gnonspec': [0.0], 'ik': [0.0], 'minf': [0.0], 'ninf': [0.0], 'uinf': [0.0], 'mtau': [0.0], 'ntau': [0.0], 'utau': [0.0], 'ino': [0.0], 'm': [0.0], 'n': [0.0], 'u': [0.0]}, 'glia__dbbs_mod_collection__Kv3_4__0': {'gkbar': [0.00076192450952], 'ik': [0.0], 'minf': [0.0], 'mtau': [0.0], 'hinf': [0.0], 'htau': [0.0], 'm': [0.0], 'h': [0.0]}}, 'ions': {'na': {'ena': [50.0], 'nai': [10.0], 'nao': [140.0], 'ina': [0.0], 'dina_dv_': [0.0]}, 'k': {'ek': [-77.0], 'ki': [54.4], 'ko': [2.5], 'ik': [0.0], 'dik_dv_': [0.0]}, 'ca': {'eca': [132.4579341637009], 'cai': [5e-05], 'cao': [2.0], 'ica': [0.0], 'dica_dv_': [0.0]}, 'no': {'eno': [0.0], 'noi': [1.0], 'noo': [1.0], 'ino': [0.0], 'dino_dv_': [0.0]}}, 'morphology': {'L': 5.622320175170898, 'diam': [5.800000190734863], 'pts3d': [(0.0, 0.0, 0.0, 5.800000190734863), (0.0, 5.622320175170898, 0.0, 5.800000190734863)], 'parent': None, 'trueparent': None}, 'nseg': 1, 'Ra': 100.0, 'cm': [1.0], 'regions': set(), 'species': set(), 'name': 'SBDEAEIOZX_1_interval (0.0-1)', 'hoc_internal_name': '__nrnsec_0x555681876560', 'cell': None}
```

But this does not really match what I asked in the definition. For instance I defined the cm as 2, but it was actually set 1. Something strange occurs for ek, and probably for other parameters. What is wrong with my definition?

**Comment by dariosannino** [comment]
PS. Mechanims introduced in the definition seem to be set properly.
What is defined in ```cable``` and ```ions``` not.

**Comment by dariosannino** [comment]
After you solved the problem, my granule cell works fine now and gets updated properly.
However I found some problems with the Purkinje cell. Now, I am not able anymore to build the cell. I will give you the definition and the commands that I am writing in the CLI:

```from arborize import define_model

definitionPCWT = define_model({
    "synapse_types": {
        "AMPA": {
            "mechanism": "AMPA",
            "parameters": {
                "tau_facil": 54,
                "tau_rec": 35.1,
                "tau_1": 6,
                "gmax": 1200,
                "U": 0.13,
            },
        },
        "GABA": {
            "mechanism": "GABA",
            "parameters": {
                "tau_facil": 4,
                "tau_rec": 15,
                "tau_1": 1,
                "Erev": -70,
                "gmax": 2600,
                "U": 0.35,
            },
        },
    },
    "cable_types": {
        "soma": {
            "cable": {"Ra": 122, "cm": 1},
            "ions": {
                "na": {"rev_pot": 60},
                "k": {"rev_pot": -88},
                "h": {"rev_pot": -34.4},
                "ca": {"rev_pot": 137.52625}
                },
            "mechanisms": {
                "Leak": {"e": -61, "gmax": 0.001},
                "Nav1_6": {"gbar": 0.19835422387652},
                "Kv1_1": {"gbar": 0.00185494695569},
                "Kv3_4": {"gkbar": 0.05684366437809},
                "Kir2_3": {"gkbar": 2.100544604e-05},
                "Cav2_1": {"pcabar": 0.0001530123689},
                "Cav3_1": {"pcabar": 4.19659246e-06},
                "Cav3_2": {"gcabar": 0.00067268089296},
                "Cav3_3": {"pcabar": 0.00015401124325},
                "Kca1_1": {"gbar": 0.00877088561313},
                "Kca2_2": {"gkbar": 0.00085748826176},
                "Kca3_1": {"gkbar": 0.00978230722317},
                "HCN1": {"gbar": 0.00192222696826},
                ("cdp5", "CAM"): {"TotalPump": 2e-08},
            },
            "synapses": {
                "GABA": {
                    "mechanism": "GABA",
                    "parameters": {
                        "tau_facil": 4,
                        "tau_rec": 15,
                        "tau_1": 1,
                        "Erev": -70,
                        "gmax": 2600,
                        "U": 0.35,
                    },
                },
            },
        },
        "dendrites": {
            "cable": {
                # "cm": lambda d: (11.510294 * math.exp(-1.376463 * d) + 2.120503),
                # Formula for cm varied from 2 to 6, aprroximated with avg
                "cm": 5.577701084442752,
                "Ra": 122,
            },
            "ions": {
                "na": {"rev_pot": 60},
                "k": {"rev_pot": -88},
                "h": {"rev_pot": -34.4},
                "ca": {"rev_pot": 137.52625},
            },
            "mechanisms": {
                "Leak": {"e": -61, "gmax": 0.0003},
                "Kv1_1": {"gbar": 0.00105296938319},
                "Kv1_5": {"gKur": 0.00018237109331},
                "Kv3_3": {"gbar": 0.01192448545734},
                "Kv4_3": {"gkbar": 0.00082020749325},
                "Cav2_1": {"pcabar": 0.001594103024736},
                "Cav3_3": {"pcabar": 0.00013000262047},
                "Kca1_1": {"gbar": 0.03803999223084},
                "HCN1": {"gbar": 2.49811225e-06},
                ("cdp5", "CAM"): {"TotalPump": 5e-08},
            },
            "synapses": {
                "AMPA": {
                    "mechanism": "AMPA",
                    "parameters": {
                        "tau_facil": 54,
                        "tau_rec": 35.1,
                        "tau_1": 6,
                        "gmax": 1200,
                        "U": 0.13,
                    },
                },
            },
        },
        "basal_dendrites": {
            "cable": {"Ra": 122, "cm": 1},
            "ions": {
                "na": {"rev_pot": 60}, "k": {"rev_pot": -88}, "ca": {"rev_pot": 137.52625}
                },
            "mechanisms": {
                "Leak": {"e": -61, "gmax": 0.0003},
                "Kir2_3": {"gkbar": 1.135399111e-05},
                "Cav3_1": {"pcabar": 4.0066819e-06},
                "Cav3_2": {"gcabar": 0.00194175819441},
                "Kca2_2": {"gkbar": 0.00078534639818},
                "Kca3_1": {"gkbar": 0.00330481139341},
            },
        },
        "sodium_dendrites": {
            "cable": {"Ra": 122, "cm": 1},
            "ions": {
                "na": {"rev_pot": 60}, "k": {"rev_pot": -88}, "ca": {"rev_pot": 137.52625}
                },
            "mechanisms": {
                "Leak": {"e": -61, "gmax": 0.0003},
                "Kir2_3": {"gkbar": 1.135399111e-05},
                "Cav3_1": {"pcabar": 4.0066819e-06},
                "Cav3_2": {"gcabar": 0.00194175819441},
                "Kca2_2": {"gkbar": 0.00078534639818},
                "Kca3_1": {"gkbar": 0.00330481139341},
                "Nav1_6": {"gbar": 0.01568012827236},
            },
        },
        "aa_targets": {"cable": {}, "ions": {}, "mechanisms": {}},
        "pf_targets": {"cable": {}, "ions": {}, "mechanisms": {}},
        "sc_targets": {"cable": {}, "ions": {}, "mechanisms": {}},
        "AIS": {
            "cable": {"Ra": 122, "cm": 1},
            "ions": {
                "na": {"rev_pot": 60}, "k": {"rev_pot": -88}, "ca": {"rev_pot": 137.52625}
                },
            "mechanisms": {
                "Leak": {"e": -61, "gmax": 0.0003},
                "Nav1_6": {"gbar": 0.76520432560779},
                "Kv3_4": {"gkbar": 0.01470016164534},
                "Cav2_1": {"pcabar": 0.00028786836482},
                "Cav3_1": {"pcabar": 6.43429659e-06},
                ("cdp5", "CAM"): {"TotalPump": 2e-08},
            },
        },
        "axon": {"cable": {}, "ions": {}, "mechanisms": {}},
        "AIS_K": {
            "cable": {"Ra": 122, "cm": 1},
            "ions": {
                "k": {"rev_pot": -88}
                },
            "mechanisms": {
                "Leak": {"e": -61, "gmax": 0.0003},
                "Kv1_1": {"gbar": 0.01011038066089},
            },
        },
        "axonmyelin": {
            "cable": {"Ra": 122, "cm": 1.87e-11},
            "ions": {},
            "mechanisms": {"pas": {"e": -61, "g": 5.6e-09}},
        },
        "nodes": {
            "cable": {"Ra": 122, "cm": 1},
            "ions": {
                "na": {"rev_pot": 60}, "k": {"rev_pot": -88}, "ca": {"rev_pot": 137.52625}
                },
            "mechanisms": {
                "Leak": {"e": -61, "gmax": 0.0003},
                "Nav1_6": {"gbar": 0.02749309338198},
                "Kv3_4": {"gkbar": 0.02227585026371},
                "Cav2_1": {"pcabar": 0.00012353396215},
                "Cav3_1": {"pcabar": 1.508814156e-05},
                ("cdp5", "CAM"): {"TotalPump": 5e-07},
            },
        },
    },
})


tagsPC = {
    16: ["axon", "AIS"],
    17: ["axon", "AIS_K"],
    18: ["axon", "axonmyelin"],
    19: ["axon", "nodes"],
    20: ["dendrites", "basal_dendrites"],
    21: ["dendrites", "pf_targets", "sc_targets"],
    22: ["dendrites", "aa_targets", "sc_targets"],
}
```

After the definition I continue in this way:
```
morpho = Morphology.from_swc("Morphologies/PurkinjeCell.swc", tags=tagsPC)
schematic = bsb_schematic(morpho, definitionPCWT)
cell = neuron_build(schematic)
```

And then it returns this error:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/mnt/c/Users/pc/.vscode/Thesis_2/arborize/arborize/builders/_neuron.py", line 131, in neuron_build
    schematic.freeze()
  File "/mnt/c/Users/pc/.vscode/Thesis_2/arborize/arborize/schematic.py", line 128, in freeze
    self._flatten_branches(self.roots)
  File "/mnt/c/Users/pc/.vscode/Thesis_2/arborize/arborize/schematic.py", line 147, in _flatten_branches
    self._flatten_branches(branch.children)
  File "/mnt/c/Users/pc/.vscode/Thesis_2/arborize/arborize/schematic.py", line 147, in _flatten_branches
    self._flatten_branches(branch.children)
  File "/mnt/c/Users/pc/.vscode/Thesis_2/arborize/arborize/schematic.py", line 134, in _flatten_branches
    branch.definition = self._makedef(branch.labels)
  File "/mnt/c/Users/pc/.vscode/Thesis_2/arborize/arborize/schematic.py", line 160, in _makedef
    return CableType.anchor(
  File "/mnt/c/Users/pc/.vscode/Thesis_2/arborize/arborize/definitions.py", line 142, in anchor
    def_.merge(def_right)
  File "/mnt/c/Users/pc/.vscode/Thesis_2/arborize/arborize/definitions.py", line 147, in merge
    self._mergedict(self.ions, def_right.ions)
  File "/mnt/c/Users/pc/.vscode/Thesis_2/arborize/arborize/definitions.py", line 154, in _mergedict
    dself[key].merge(dother[key])
AttributeError: 'Ion' object has no attribute 'merge' 
```

Doing a bit of debugging it seems that the error stands in "dendrites". Its ion description is identical to the one of the "soma", but the one of the dendrites give problems. I discovered this removing/commenting the ions in each section, and the only one that gives problems is that of "dendrites". Hope it could help. If there is somenthing wrong in my code let me know, please.
Thank you.

**Comment by dariosannino** [comment]
Fixed in c920308933d3e69a98893af381f4f4978db43343
